File: src/others/4-median-of-2-sorted-array/typescript/median_of_2_sorted_array.ts
export function findMedianSortedArrays_mergeSort(
  nums1: number[],
  nums2: number[],
): number {
  const m = nums1.length,
    n = nums2.length;
  let p1 = 0,
    p2 = 0;

  const getMin = function (): number {
    if (p1 < m && p2 < n) {
      return nums1[p1] < nums2[p2] ? nums1[p1++] : nums2[p2++];
    } else if (p1 < m) {
      return nums1[p1++];
    } else if (p2 < n) {
      return nums2[p2++];
    }
    return -1;
  };

  if ((m + n) % 2 === 0) {
    for (let i = 0; i < (m + n) / 2 - 1; ++i) {
      getMin();
    }
    return (getMin() + getMin()) / 2;
  } else {
    for (let i = 0; i < Math.floor((m + n) / 2); ++i) {
      getMin();
    }
    return getMin();
  }
}

export function findMedianSortedArrays_binarySearch(
  nums1: number[],
  nums2: number[],
): number {
  if (nums1.length > nums2.length) {
    const temp = nums1;
    nums1 = nums2;
    nums2 = temp;
  }
  const m = nums1.length,
    n = nums2.length;
  let left = 0,
    right = m;
  while (left <= right) {
    const partitionA = Math.floor((left + right) / 2);
    const partitionB = Math.floor((m + n + 1) / 2 - partitionA);
    const maxLeftA =
      partitionA == 0 ? Number.MIN_SAFE_INTEGER : nums1[partitionA - 1];
    const minRightA =
      partitionA == m ? Number.MAX_SAFE_INTEGER : nums1[partitionA];
    const maxLeftB =
      partitionB == 0 ? Number.MIN_SAFE_INTEGER : nums2[partitionB - 1];
    const minRightB =
      partitionB == n ? Number.MAX_SAFE_INTEGER : nums2[partitionB];
    if (maxLeftA <= minRightB && maxLeftB <= minRightA) {
      if ((m + n) % 2 == 0) {
        return (
          (Math.max(maxLeftA, maxLeftB) + Math.min(minRightA, minRightB)) / 2.0
        );
      } else {
        return Math.max(maxLeftA, maxLeftB);
      }
    } else if (maxLeftA > minRightB) {
      right = partitionA - 1;
    } else {
      left = partitionA + 1;
    }
  }
  return 0;
}

File: src/others/4-median-of-2-sorted-array/cpp/src/median_of_2_sorted_array.cpp
#include "median_of_2_sorted_array.hpp"

#include <climits>

using namespace std;

// Get the smaller value between nums1[p1] and nums2[p2] and move the
// pointer forward.
Solutions::Solutions() : p1(0), p2(0)
{
}
int Solutions::getMin(const vector<int> &nums1, const vector<int> &nums2)
{
    if (p1 < nums1.size() && p2 < nums2.size())
    {
        return nums1[p1] < nums2[p2] ? nums1[p1++] : nums2[p2++];
    }
    else if (p1 < nums1.size())
    {
        return nums1[p1++];
    }
    else if (p2 < nums2.size())
    {
        return nums2[p2++];
    }
    return -1;
}

float Solutions::findMedianSortedArrays_mergeSort(const vector<int> &nums1, const vector<int> &nums2)
{
    int m = int(nums1.size()), n = int(nums2.size());

    if ((m + n) % 2 == 0)
    {
        for (int i = 0; i < (m + n) / 2 - 1; ++i)
        {
            int _ = getMin(nums1, nums2);
        }
        int a = getMin(nums1, nums2);
        int b = getMin(nums1, nums2);
        return (float)(a + b) / 2;
    }
    else
    {
        for (int i = 0; i < (m + n) / 2; ++i)
        {
            int _ = getMin(nums1, nums2);
        }
        return getMin(nums1, nums2);
    }

    return -1;
}

float Solutions::findMedianSortedArrays_binarySearch(const vector<int> &nums1, const vector<int> &nums2)
{
    if (nums1.size() > nums2.size())
    {
        return findMedianSortedArrays_binarySearch(nums2, nums1);
    }

    int m = nums1.size(), n = nums2.size();
    int left = 0, right = m;

    while (left <= right)
    {
        int partitionA = (left + right) / 2;
        int partitionB = (m + n + 1) / 2 - partitionA;

        int maxLeftA = (partitionA == 0) ? INT_MIN : nums1[partitionA - 1];
        int minRightA = (partitionA == m) ? INT_MAX : nums1[partitionA];
        int maxLeftB = (partitionB == 0) ? INT_MIN : nums2[partitionB - 1];
        int minRightB = (partitionB == n) ? INT_MAX : nums2[partitionB];

        if (maxLeftA <= minRightB && maxLeftB <= minRightA)
        {
            if ((m + n) % 2 == 0)
            {
                return (max(maxLeftA, maxLeftB) +
                        min(minRightA, minRightB)) /
                       2.0;
            }
            else
            {
                return max(maxLeftA, maxLeftB);
            }
        }
        else if (maxLeftA > minRightB)
        {
            right = partitionA - 1;
        }
        else
        {
            left = partitionA + 1;
        }
    }

    return 0.0;
}
